### PR TITLE
add OsDiskType configuration for AKS node pools

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR enables OSDiskType configuration for AKS cluster node pool (`AzureManagedMachinePool`) resources.

This is an optional configuration for `AzureManagedMachinePool ` resources, and validation will be enforce to ensure that only valid values are allowed: "Managed" or "Ephemeral".

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
add OSDiskType configuration for AKS node pools
```

### CODE REVIEW

File: templates/flavors/aks/cluster-template.yaml
Comment (jackfrancis) [inline templates/flavors/aks/cluster-template.yaml:77]: These are temporary changes so that we can run tests before merging. Once we have passing signal I'll remove them.

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:218]: This (and the updates to UT) will be cleaned up once https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1915 merges

**Comment by CecileRobertMichon** [comment]
#1915 merged

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: Unnecessary foo no longer present

File: templates/test/ci/cluster-template-prow-aks-multi-tenancy.yaml
Comment (jackfrancis) [inline templates/test/ci/cluster-template-prow-aks-multi-tenancy.yaml:105]: Since we're adding a 3rd pool to test the "no configuration specified" pool config scenario, we may as well add a differnt `maxPods` config value to the 2nd one that now serves to validate the 2nd possible valid `OsDiskType` flavor.

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (CecileRobertMichon) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:230]: Forgot to mention this before for maxPods validation but we should be using api-machinery field validation errors for these (similar to what we do above with field.Invalid). Error types are found here: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/field/errors.go#L93

`field.NotSupported()` seems like a good fit for this one

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: Done.

Addressed the MaxPods implementation with this PR:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1920

File: azure/scope/managedcontrolplane_test.go
Comment (zmalik) [inline azure/scope/managedcontrolplane_test.go:296]:
```suggestion
				PatchTarget:      getAzureMachinePoolWithOsDiskType("pool1", string(containerservice.OSDiskTypeEphemeral)),,
```

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (zmalik) [inline exp/api/v1beta1/azuremanagedmachinepool_types.go:73]: lets add the openAPI schema validation for this field
// +kubebuilder:validation:Enum= Ephemeral; Managed

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (zmalik) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:228]: that can be delegated to openAPI schema validation

File: azure/scope/managedcontrolplane_test.go
Comment (jackfrancis) [reply azure/scope/managedcontrolplane_test.go:296]: yikes, good catch, I wish the UT result would not have permitted that...

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:228]: @CecileRobertMichon do you agree this webhook validation is superfluous if we add the enum allowed values to the api property spec?

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (zmalik) [inline exp/api/v1beta1/azuremanagedmachinepool_types.go:76]: should we use a pointer here to make it truly optional? 
and if it's not optional in sense, there should be always some default value can we add this to mutation webhook?

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (CecileRobertMichon) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:231]: +1, for simple enumeration validation, it's better to use openAPI schema validation and only use webhook validation for more complex validation.

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (CecileRobertMichon) [reply exp/api/v1beta1/azuremanagedmachinepool_types.go:76]: +1 

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (AverageMarcus) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:154]: Do we need to handle both situations individually? Can they not be combined into a single "immutable" error?

```suggestion
	if old.Spec.OsDiskType != "" && r.Spec.OsDiskType != old.Spec.OsDiskType  {
		// Prevent OSDiskType modification if it was already set to some value
		allErrs = append(allErrs,
			field.Invalid(
				field.NewPath("Spec", "OsDiskType"),
				r.Spec.OsDiskType,
				"field is immutable"))
	}

```

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:154]: I think your observation here is sound. If `A` is a non-empty string, then we can simply test `B` for being different from `A`. `B` being an empty string will always be different from `A` when `A` is a non-empty string, so we can exclusively use:

- `if r.Spec.OsDiskType != old.Spec.OsDiskType`

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_types.go]: done

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: done

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: Because other validations in this surface area are doing this "test for empty string" validation, let's continue this pattern for now. If we want to change this approach let's get consensus and then change it everywhere so that we're consistent.

Thoughts?

**Comment by jackfrancis** [comment]
@CecileRobertMichon @zmalik I think this is ready for another review

File: azure/scope/managedcontrolplane.go
Comment (jackfrancis) [inline azure/scope/managedcontrolplane.go:504]: @zmalik @CecileRobertMichon moved these assignments here (instead of lower in a "if not empty" condition) in response to this PR:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1945

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (AverageMarcus) [inline exp/api/v1beta1/azuremanagedmachinepool_types.go:74]:
```suggestion
	// +kubebuilder:validation:Enum=Ephemeral;Managed
```

Nitpicking as I don't _think_ it actually matters but for consistency sake :)

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (AverageMarcus) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:47]: Could this be handled using a default comment on the type for OpenAPI to handle for us? (I'm not actually sure if that would handle both nil and empty string)

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (AverageMarcus) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: +1 for consistency. I hadn't noticed the similar pattern above.

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (AverageMarcus) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:233]: Do we still need this function?

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:233]: I have a slight preference towards shipping a new feature w/ a "prototype" validation function to standardize how we maintain these validations (so that if we add validations over time they are structured/expressed in a consistent way across the codebase).

@CecileRobertMichon thoughts?

Comment (CecileRobertMichon) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:233]: I'm leaning towards not keeping a function that's a no-op / not needed. If we come across the need for more validation that can't be done in OpenAPI later we can easily add this back (there are plenty of other examples of similar validation functions in the codebase)

See also #1338

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:47]: I think we could add this:

`// +kubebuilder:default=Managed`

@CecileRobertMichon @zmalik is that preferable to enforcing in the `Default()` method of the webhook?

Comment (CecileRobertMichon) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:47]: Both are valid ways to do defaulting. I don't know if one is preferable over the other. OpenAPI defaulting is slightly cleaner since it doesn't require code in the webhook, but I don't know that it's a huge gain. It might be good to check if there are any API conventions out there that recommend one over the other (there isn't anything about this in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:47]: I would prefer using the openapi solution, I'll test drive that out.

@zmalik do you have any thoughts?

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: removed

File: exp/api/v1beta1/azuremanagedmachinepool_types.go
Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_types.go]: fixed, thanks!

**Comment by jackfrancis** [comment]
@AverageMarcus @zmalik @CecileRobertMichon this is ready for another (possibly final) round of reviews, thank you!

**Review by AverageMarcus — APPROVED** [review]
LGTM 👍

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (CecileRobertMichon) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:48]: It looks like you added a default in the OpenAPI schema, can we remove this?

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:48]: Yes, thanks!

Because the schema is now injecting the default value, are we able to use validation to fail if nil or empty string?

File: templates/test/ci/prow-aks-multi-tenancy/patches/aks-agentpool0.yaml
Comment (CecileRobertMichon) [inline templates/test/ci/prow-aks-multi-tenancy/patches/aks-agentpool0.yaml:6]: let's remove this and wait for #1971 to merge

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go
Comment (zmalik) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: I also think that both ways are valid. And not sure about the standard preferred way. If it can be done through OpenAPI I would also prefer that way. 

> I'm not actually sure if that would handle both nil and empty string

from 1.20 we have this: 
```
When nullable is true, null values will be conserved and won't be defaulted
```

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: Based on prior discussions about preferring an opinionated approach vs a passive one (let AKS choose the default for us), I think we should keep the current definition and not set `OsDiskType` to nullable.

File: templates/test/ci/prow-aks-multi-tenancy/patches/aks-agentpool0.yaml
Comment (jackfrancis) [reply templates/test/ci/prow-aks-multi-tenancy/patches/aks-agentpool0.yaml]: rebased

**Review by CecileRobertMichon — COMMENTED** [review]
/approve

**Pull request merged**
